### ISSUE TITLE
Instances of DirectoryFileReference and byte[] keep being created and never cleaned up by the Garbage Collector

### ISSUE THREAD

**Issue by FollowSteph** [PathmindAI/pathmind-webapp#344]
For some reason the instance count of `DirectoryFileReference` keeps increasing and increasing and is never collected by the Garbage collector. This is a memory leak that will eventually cause the system to run out of memory.

I'm not sure if this is related but there a hotspot in the code related to `Flux`. With that in mind I see `RescaleRestApiClient.workingFiles()` using a combination of the two so this is where I would first investigate. 

I will update this issue with more details shortly, including specific data from the code profiler with charts and all which should give us a lot better insight.

**UPDATE**

After further investigation this only seems to happen after a discovery (and maybe full run, I didn't test that) is started. As soon as it's started instances of `DirectoryFileReference` are created and never freed. 

![image](https://user-images.githubusercontent.com/19716286/67435229-ad7e4800-f5b9-11e9-8709-e6649fbe7586.png)

Below is a chart of the heap over time. You can see a massive jump at around 25 minutes. This is when I started a discovery run. At 30 minutes is when I started to get results. The memory usage just keeps climbing and climbing from that point on. That's when instances of `DirectoryFileReference` start to appear and they just keep climbing and never being freed.

IMPORTANT: Keep in mind the system is idle and there are no interactions happening with the webapp during the recording of these metrics.

![image](https://user-images.githubusercontent.com/19716286/67435946-3fd31b80-f5bb-11e9-9871-dce968df87cd.png)

Instance count at 45 minutes:

![image](https://user-images.githubusercontent.com/19716286/67436296-0e0e8480-f5bc-11e9-9123-0d3af1f5974f.png)

And associated heapstack and memory usage at 45 minutes. Notice it's just climbing at a pretty steady pace.

![image](https://user-images.githubusercontent.com/19716286/67436341-267e9f00-f5bc-11e9-813b-c30c46ce2cbc.png)

The memory consumption kept increasing at that pace for about an hour after which point the slope decreased, that is to say the memory consumption kept increasing just not as quickly.

**UPDATE**

This is the heap chart after almost 400 minutes:

![image](https://user-images.githubusercontent.com/19716286/67450526-a96b1e00-f5eb-11e9-95ce-7a8159ebf0a5.png)

You will notice that although it somewhat stabilized after the initial increase memory usage still kept creeping up. The increase speed in memory usage around 250 minutes is when I did a full run. It seems to also have a memory leak but not as pronounced as a discovery. This is confirmed with the last blip around 350 minutes at which point I ran a discovery on another new project. So after 400'ish minutes I went from 500mb usage to about 7.5gb in total memory with almost 5gb allocated to objects in the heap space.

At this point `DirectoryFileReference` consumes the most memory after `byte[]` and `String` (`byte[]` and `String` generally consumer the most memory). What is odd though is how many `HashMap` entries that are still held in memory. I'm wondering if `DirectoryFileReference` is not being held as a reference in a `HashMap`. 

That being said the CPU hotspot seems to be elsewhere (which is a separate issue #350):

![image](https://user-images.githubusercontent.com/19716286/67450797-8b51ed80-f5ec-11e9-8e06-9d87cfda250c.png)

In any case basically over time more and more memory is being allocated until it consumes the server and we run into a heap space error that we ran into today which requires a server reboot.

**IMPORTANT** These metrics are from my own personal branch which has a lot of the performance improvements for #312 and so you may not experience the same results until I commit my fixes for #312 and they are merged into the dev branch. That is to say you will see the compound of both issues rather than just this one as I am seeing in the above metrics. 

**UDPATE** To follow up here's another screenshot showing the number of instances of DirectoryFileReference after #379 and the memory improvements. This is after running several discoveries and full runs back to back. You can see the instance count is quite high. And once the runs were completed the instance count seems to stay, so for whatever reason they are not being released from memory.

![image](https://user-images.githubusercontent.com/19716286/68565964-10c60200-0423-11ea-8d12-58eca56e1e7a.png)
